trl/trainer/dpo_trainer.py

```diff
@@ -1478,67 +1314,42 @@ def get_batch_loss_metrics(
         """Compute the DPO loss and other metrics for the given batch of inputs for train or test."""
         metrics = {}
 
-        forward_output = self.concatenated_forward(model, batch)
-        (
-            policy_chosen_logps,
-            policy_rejected_logps,
-            policy_chosen_logits,
-            policy_rejected_logits,
-            policy_nll_loss,
-            policy_weights,
-        ) = forward_output[:6]
-        if self.aux_loss_enabled:
-            aux_loss = forward_output[6]
+        model_output = self.concatenated_forward(model, batch)
 
         # if reference_chosen_logps and reference_rejected_logps in batch use them, otherwise use the reference model
-        if (
-            "reference_chosen_logps" in batch
-            and "reference_rejected_logps" in batch
-            and (self.precompute_ref_log_probs or self.args.rpo_alpha is not None)
-        ):
-            reference_chosen_logps = batch["reference_chosen_logps"]
-            reference_rejected_logps = batch["reference_rejected_logps"]
+        if "reference_chosen_logps" in batch and "reference_rejected_logps" in batch:
+            ref_chosen_logps = batch["reference_chosen_logps"]
+            ref_rejected_logps = batch["reference_rejected_logps"]
         else:
-            with torch.no_grad():
-                if self.ref_model is None:
-                    with self.null_ref_context():
-                        reference_chosen_logps, reference_rejected_logps = self.concatenated_forward(
-                            self.model, batch
-                        )[:2]
-                else:
-                    reference_chosen_logps, reference_rejected_logps = self.concatenated_forward(
-                        self.ref_model, batch
-                    )[:2]
+            ref_chosen_logps, ref_rejected_logps = self.compute_reference_log_probs(batch)
 
         losses, chosen_rewards, rejected_rewards = self.dpo_loss(
-            policy_chosen_logps,
-            policy_rejected_logps,
-            reference_chosen_logps,
-            reference_rejected_logps,
+            model_output["chosen_logps"], model_output["rejected_logps"], ref_chosen_logps, ref_rejected_logps
         )
         reward_accuracies = (chosen_rewards > rejected_rewards).float()
 
         if self.args.rpo_alpha is not None:
-            # RPO loss from V3 of the paper:
-            losses = losses + policy_nll_loss * self.args.rpo_alpha
+            losses = losses + self.args.rpo_alpha * model_output["nll_loss"]  # RPO loss from V3 of the paper
 
         if self.use_weighting:
-            losses = losses * policy_weights
+            losses = losses * model_output["policy_weights"]
```

Comment:
Answering https://github.com/huggingface/trl/pull/2209/files/9824f1670bb08845ab7bb850b5b14671bf843da5#r1804420747

I'm not 100% sure to understand your question. When the weighting is disabled, the loss won't be multiplied by the weights here. It's not even in `model_output`.
Please tell me if I've missed something.